File: 141_linked_list_cycle/two_pointers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == NULL)
            return false;
            
        else if (head->next == NULL)
            return false;

        bool hasCycle = false;
        ListNode* ptr;
        ListNode* ptr_skip;
        ptr = head;
        ptr_skip = head->next;

        while (!hasCycle && ptr_skip != NULL)
        {
            if(ptr == ptr_skip)
                hasCycle = true;
            ptr = ptr->next;
            if (ptr_skip->next == NULL)
                break;
            ptr_skip = ptr_skip->next->next;
        }

        return hasCycle;  
    }
};

File: 141_linked_list_cycle/two_pointers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None:
            return False
        
        ptr = head
        ptr_skip = head.next

        while ptr_skip != None:
            if ptr == ptr_skip:
                return True

            ptr = ptr.next
            if ptr_skip.next == None:
                break
            
            ptr_skip = ptr_skip.next.next

        return False